### DISCUSSION TITLE
Different Time Recorded for Allreduce Operation

### DISCUSSION THREAD

**Discussion by dexhunter** [horovod/horovod#3067] [Q&A]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) Pytorch
2. Framework version:  1.7.1
3. Horovod version:  0.22.0
4. MPI version: 4.0.1
5. CUDA version: 10.1
6. NCCL version: 2.6.4
7. Python version: 3.7.10
8. Spark / PySpark version:
9. Ray version:
10. OS and version: ubuntu 
11. GCC version: 5.4.0
12. CMake version: 

**Bug report:**

I use a simple script to record the elapsed time for allreduce op but I am getting different elapsed time from `torch.cuda.Event(enable_timing=True)` and the built-in timeline of horovod. 

My script
```python
import torch
import horovod.torch as hvd
import numpy as np

hvd.init()
torch.cuda.set_device(hvd.local_rank())

start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)
tensor_size = 2**28
x = torch.randn(tensor_size, dtype=torch.float).cuda()

# allreduce op
start.record()
reduced = hvd.allreduce(x, average=False)
end.record()

torch.cuda.synchronize()

elapsed_time=start.elapsed_time(end)
```

However, when I print out the result, there is always a gap between what torch records and the time on the timeline. I am just wondering which one will be more accurate? Thanks!

**Comment by dexhunter** [comment]
Another issue I met is that I am not able to get the same result from what I saw on chrome://tracing and the json file. I extract the time based on `pid` but the results are a bit off. Any helps there? Thanks in advance!

For example, I have this [json file](https://gist.github.com/dexhunter/bd1db106ba1c199cfdb65ea61667682b) from `HOROVOD_TIMELINE` and an extract time script gets `ts` based `pid`. But the results I got are different.  For `pid=28` I get `815.850`ms from python script and `818.350`ms from chrome://tracing.
![image](https://user-images.githubusercontent.com/6930518/124086216-68dfcb00-da83-11eb-92d8-14e1fd00b65e.png)

![image](https://user-images.githubusercontent.com/6930518/124086191-63828080-da83-11eb-92e7-f50d625edba1.png)
+

**Comment by maxhgerlach** [comment]
> I use a simple script to record the elapsed time for allreduce op but I am getting different elapsed time from torch.cuda.Event(enable_timing=True) and the built-in timeline of horovod.

There can be some small extra delays from framework code. I don't know too much about PyTorch, but at least with graph-based TensorFlow code these things tend to get better in subsequent training steps. How large are the discrepancies you've seen, btw?

I've found Nvidia's Nsight profiling tools really helpful to understand where my models spend time. To make this easier with Horovod, I've contributed a PR #2723 that allows you to both see (1) timing ranges approximately corresponding to your torch timings and (2) all the detailed Horovod timeline information in the same profiling display.

**Comment by dexhunter** [comment]
> How large are the discrepancies you've seen, btw?

![image](https://user-images.githubusercontent.com/6930518/124258864-7919a800-db60-11eb-8f42-b63c7f8c2370.png)

@maxhgerlach Hi Max, thanks for the reply. The horovod records `NEGOTIATE_ALLREDUCE` for ~4949ms for the discrepancy is really large for the first pid, the rest difference is a bit smaller but still about ~30% difference between time recorded by torch and horovod timeline

**Comment by maxhgerlach** [comment]
@dexhunter, on your plot the extra latency seems to grow with the message size. Might this just be latency incurred by copying around the data?

**Comment by dexhunter** [comment]
@maxhgerlach Hi thanks for the reply. From my understanding, with larger message sizes, the time for executing the operation will also grow. The plot is based on the timeline json file, so the time moving around the time should also be included as seen in WAIT_FOR_DATA or maybe I should change the ylabel to execution time?
